### ISSUE TITLE
Training the model using with a custom triplet loss function causes the weights to explode and start predicting nan values

### ISSUE THREAD

**Issue by yolob21** [krasserm/face-recognition#1]
Created a triplet-generator while viewing this facenet implementation of [triplet loss function](https://github.com/davidsandberg/facenet/blob/master/src/train_tripletloss.py). Then combined the model from the python notebook mentioned in this repo. I am training using the vgg2face dataset.

But training fails and stop after just 1 epoch step. The weights seem to be exploding,unable to figure out why. Attached the python notebook which shows the code I used for training

[project-cleaned_2.0.zip](https://github.com/krasserm/face-recognition/files/1957690/project-cleaned_2.0.zip)

**Comment by krasserm** [comment]
I didn't run your notebook but I've seen that `triplet_loss_v2` in your notebook doesn't include margin `alpha` which prevents the network from learning anything useful (and maybe even converge).

**Comment by yolob21** [comment]
So I tried with alpha as follows:

```
def triplet_loss_v2(y_true,y_preds):
    reshape_triplets_embeddings=tf.reshape(y_preds,[-1,3,128])
    an,pn,nn=tf.split(reshape_triplets_embeddings,3,1)
    a=tf.reshape(an,[-1,128])
    p=tf.reshape(pn,[-1,128])
    n=tf.reshape(nn,[-1,128])
    p_dist = K.sum(K.square(a-p), axis=-1)
    n_dist = K.sum(K.square(a-n), axis=-1)
    return K.sum(K.maximum(p_dist - n_dist+0.2, 0), axis=0)
```

The result remains the same,network predicts nan values at the same epoch step. changing the dataset incase I had any bad input also gives the same result.

In the face-recognition python notebook attached in this repo, you have added a Triplet loss Layer to a new model(#2) and then trained that model(#2) what I didnt understand was is how will the weights for the model(#1) which generate the embeddings be modified if one is training a model(#2) which only takes the input from embedding generating model(#1)

**Comment by krasserm** [comment]
Assuming you mean by model 1 `nn4_small2 ` and by model 2 `nn4_small2_train`. Model 2 in this case uses model 1 as [shared layer](https://keras.io/getting-started/functional-api-guide/#shared-layers). Take a look at `triplet_loss_layer` which takes as input the outputs of model 1 (i.e. `emb_a`, `emb_p`, `emb_n`) in this case. So when training model 2 the weights of model 1 will be updated.

**Comment by yolob21** [comment]
Thanks for the info,so i spent some time again trying to train the network using nn4_small2 as a shared layer and reached the same conclusion as using the custom loss function. The model starts predicting nan after 1 epoch step. could you maybe try running the notebook with your custom triplet_generator function to see if the problem is with the generator or the weights.

**Comment by krasserm** [comment]
- What do you mean with with "your custom triplet_generator"? The dummy triplet generator as in https://github.com/krasserm/face-recognition/blob/master/data.py.
- What do you mean with "the notebook"? The one you provided or the one in this repo?

Or maybe a more general question: what can I try running that you can't try yourself? 

When running the notebook in this repo as is with the dummy triplet generator the loss oscillates around 0.8 over all epochs i.e. no `nan`s are produced.

**Comment by shadymohsen** [comment]
Are the nn4.small2.v1 weights output of the trained model on the triplet loss function of the datasets or the trained model to get f(x)i?, and if the nn4.small2.v1 is trained on the triplet loss function how can i load the weights of model #2 in model #1? Thank you so much. @krasserm

**Comment by krasserm** [comment]
@shadymohsen 

> Are the nn4.small2.v1 weights output of the trained model on the triplet loss function of the datasets or the trained model to get f(x)i

The weights are the same for both models as `nn4_small` is shared within `nn4_small_train` and `nn4_small_train` doesn't add any additional learnable weights.

> how can i load the weights of model #2 in model #1?

When you train `nn4_small_train` with the triplet loss you can immediately use `nn4_small` for prediction afterwards (i.e. in the same way as `nn4_small2_pretrained`).

**Comment by shadymohsen** [comment]
i'm very sorry but i still didn't understand well. so know we have a model nn4_small=createmodel() this is a model which takes the images and get output which is the f(x) right?
another thing is that we have a pre trained model which is trained on the faces of people which is the nn4.small2.v1 this is trained to get the output which is the triplet loss of three images input which used the nn4_small=createmodel(). right?
now how can we upload the weights of the pre trained model which is already trained on three images as an input and the triplet loss as an output to the model which its output is f(x) and has 1 input image. 
also, when i input my images of my library to the model how does it know the anchor from the positive and the negative? 
thank you so much in advance :)

**Comment by krasserm** [comment]
> i'm very sorry but i still didn't understand well. so know we have a model nn4_small=createmodel() this is a model which takes the images and get output which is the f(x) right?

Yes. The same is true for `nn4_small2_pretrained` which the model instance that is used in the remainder of the notebook. `nn4_small2` is only there to demonstrate how it can be trained in context of `nn4_small2_train`.

> another thing is that we have a pre trained model which is trained on the faces of people which is the nn4.small2.v1 this is trained to get the output which is the triplet loss of three images input which used the nn4_small=createmodel(). right?

`nn4.small2.v1.h5` are only the weights that can be loaded into a model returned by `create_model()`. The notebook loads it into `nn4_small2_pretrained`. The training model `nn4_small2_train` is setup such that it shares the weights with `nn4_small`. `nn4_small2_train` itself doesn't add any further weights. Hence, the learned weights can be saved and loaded into a model instance returned by `create_model()`.

>  now how can we upload the weights of the pre trained model which is already trained on three images as an input and the triplet loss as an output to the model which its output is f(x) and has 1 input image.

See previous answer. Key is that `nn4_small2_train` and `nn4_small2` share weights.

> also, when i input my images of my library to the model how does it know the anchor from the positive and the negative?

Only the training model `nn4_small_train` needs to discriminate them. For that purpose it has three inputs `in_a`, `in_p`, `in_n`.

Hope that helps.

**Comment by snehasahu2626** [comment]
hello I have read your blog . which is really grate
but in the end i cant understand . You pass the image as 
with some number 29 and it gives the output as some person .. but whats that number 29 can you plzzz tell

**Comment by Dhruv2012** [comment]
> > i'm very sorry but i still didn't understand well. so know we have a model nn4_small=createmodel() this is a model which takes the images and get output which is the f(x) right?
> 
> Yes. The same is true for `nn4_small2_pretrained` which the model instance that is used in the remainder of the notebook. `nn4_small2` is only there to demonstrate how it can be trained in context of `nn4_small2_train`.
> 
> > another thing is that we have a pre trained model which is trained on the faces of people which is the nn4.small2.v1 this is trained to get the output which is the triplet loss of three images input which used the nn4_small=createmodel(). right?
> 
> `nn4.small2.v1.h5` are only the weights that can be loaded into a model returned by `create_model()`. The notebook loads it into `nn4_small2_pretrained`. The training model `nn4_small2_train` is setup such that it shares the weights with `nn4_small`. `nn4_small2_train` itself doesn't add any further weights. Hence, the learned weights can be saved and loaded into a model instance returned by `create_model()`.
> 
> > now how can we upload the weights of the pre trained model which is already trained on three images as an input and the triplet loss as an output to the model which its output is f(x) and has 1 input image.
> 
> See previous answer. Key is that `nn4_small2_train` and `nn4_small2` share weights.
> 
> > also, when i input my images of my library to the model how does it know the anchor from the positive and the negative?
> 
> Only the training model `nn4_small_train` needs to discriminate them. For that purpose, it has three inputs `in_a`, `in_p`, `in_n`.
> 
> Hope that helps.

![image](https://user-images.githubusercontent.com/30570613/60249111-b4ce6700-98e1-11e9-8901-a4f459484243.png)



Gives some kinda error saying weights are of model #2(nn4_small_train)  and cannot be loaded into nn4_small as it is a layer of nn4_small_train. Don't know how to fix it. If you can help


This is the code link:
Code:-  [https://github.com/Dhruv2012/Deep-Learning/blob/Align(by-Tess)/facenet/CustomModel/main.py](url)

Else portion refers Training the model and saving the weights.
If portion refers running the model and running it.
